File: Academics/BFS/8PuzzleBFS.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <stdbool.h>
#include <unistd.h>
#include "../Modules/8PuzzleModules.c"

// Not many benefits to split queue and visited memory unless the problem has step > ~17 steps
#define MAXNODES 50000 
#define HorizontalRule 45

// Forward declaration of the structure
typedef struct queueNode queueNode;

typedef struct queueNode {
    gameState State;
    queueNode *Previous;
    const char *Move; // Track the move made to reach this state
    int Depth;
} queueNode;

int Front = 0, Rear = -1, vIndex = 0;

void printHorizontalRule(){
    printf("\n");
    for (int i=0; i < HorizontalRule; i++) {
        printf("[]");
    };
}

int isQueueEmpty(queueNode Queue[]) {
    if (Front > Rear) {
        Front = 0;Rear = -1;
        printf("\n>>>[ERR 111] Queue Empty");
        return 1;
    }
    return 0;
}

int isQueueFull(queueNode Queue[]) {
    if ((Rear - Front +1) == MAXNODES) {
        printf("\n>>>[ERR 112] Queue Full");
        return 1;
    }
    return 0;
}

int isVisitedFull(queueNode Queue[]) {
    return vIndex==MAXNODES-1;
}

int pushVisited(queueNode Visited[], queueNode visitedNode) {
    if(!isVisitedFull(Visited)) {
        memcpy(&Visited[vIndex], &visitedNode, sizeof(visitedNode));
        ++vIndex;
        return 1;
    } 
    printf("\n>>>[ERR 121] Visited list Full");
    return 0;
}

int enQueue(queueNode Queue[], queueNode Node) {
    if (!isQueueFull(Queue)) {
        ++Rear;
        int Rear_Conv = Rear%MAXNODES;
        memcpy(&Queue[Rear_Conv], &Node, sizeof(Node));
        return 1;
    }
    return 0;
}

queueNode deQueue(queueNode Queue[])  {
    if(!isQueueEmpty(Queue)) {
        ++Front;
        return Queue[Front-1];
    }
}

int isVisited(gameState currentState, queueNode Visited[MAXNODES]) {
    for (int i = 0; i < vIndex; i++) {
        if (isSameState(currentState, Visited[i].State)) {
            return 1;
        }
    }
    return 0;
}

int isInStack(gameState currentState, queueNode Queue[MAXNODES]) {
    for (int i=Front; i < Rear; i++) {
        if (isSameState(currentState, Queue[i%MAXNODES].State)) {
            return 1;
        }
    }
    return 0;
}

void backtrace(queueNode *node) {
    if (node == NULL) return;
    backtrace(node->Previous);
    char formatString[50]; // Adjust size as needed
    sprintf(formatString, "Step %d [Move %s]", node->Depth, node->Move);
    printBoard(node->State, formatString); // Pass formatted string to printBoard
}

void printProgressBar(int Value1, int MAX1, int Value2, int MAX2, int period) {
    printf("\r");   // Carriage return to go back to the start of the line
    int width = 20;
    int percentage1 = Value1*100/MAX1,percentage2 = Value2*100/MAX2;
    int pos1 = (percentage1 * width)/100; 
    int pos2 = (percentage2 * width)/100;
    printf("     STATUS: Stack: ["); 
    for (int i = 0; i < width; i++) {
        if (i < pos1)
            printf("#");
        else
            printf(" ");
    }
    printf("] %d", Value1); 
    printf("    Visited: ["); 
    for (int i = 0; i < width; i++) {
        if (i < pos2)
            printf("#");
        else
            printf(" ");
    }
    printf("] %d", Value2); 
    fflush(stdout); // Flush th1e output to ensure it prints immediately
    usleep(period);
}

void solvePuzzle(gameState Initial, gameState Goal) {
    queueNode Queue[MAXNODES];
    queueNode Visited[MAXNODES]; // array
    queueNode Node;

    // Enqueue Initial State
    Node.State = Initial;
    Node.Previous = NULL;
    Node.Depth = 0;
    Node.Move = "Initial State Inclusion"; // Initial state has no move
    enQueue(Queue, Node);

    // Iteration over Queue
    printf("\n");
    while (!isQueueFull(Queue) && !isVisitedFull(Visited)) { // Added condition to avoid overflow
        printProgressBar(Rear-Front+1, MAXNODES, vIndex, MAXNODES,10000);
        // Dequeue from the Queue
        Node = deQueue(Queue);

        gameState currentState = Node.State;

        // Add to Visited Array
        if(!isVisited(currentState, Visited)) {
            pushVisited(Visited,Node);
        }

        int Depth = Node.Depth;

        // Check if currentState state is the goal state
        if (isSameState(currentState, Goal)) {
            printHorizontalRule();
            printf("\nBacktraced Path:\n");
            backtrace(&Visited[vIndex-1]);
            printf("\nSolution Found at step %d\n\n", Depth);
            return;
        }

        // Find blank position
        int x = currentState.Blank.row, y = currentState.Blank.column;

        // Possible Movements: left, right, up, down
        int dy[] = {-1, 1, 0, 0};
        int dx[] = {0, 0, -1, 1};
        const char *moves[] = {"Left", "Right", "Up", "Down"};

        for (int i = 0; i < 4; i++) {
            int newx = x + dx[i], newy = y + dy[i];

            if (isSafe(newx, newy)) {
                // Create a new State by moving blanks
                gameState newState = currentState;
                Swap(&newState.Board[x][y], &newState.Board[newx][newy]);
                newState.Blank.row = newx;
                newState.Blank.column = newy;

                // Check if the state has been visited
                if (!isVisited(newState, Visited) && !isInStack(newState,Queue)) {
                    Node.State = newState;
                    Node.Depth = Depth + 1;
                    Node.Previous = &Visited[vIndex - 1];
                    Node.Move = moves[i]; // Track the move
                    enQueue(Queue, Node);
                }
            }
        }
    }
    // No solution found within limits
    printf("\n>>>[ERR 1XX] Solution Not Found within Step Limits.\n\n");
}

int main() {
    system("clear");
    gameState Initial, Goal;

    // Inputs
    inputBoard(&Initial, "Initial");
    system("clear");
    printBoard(Initial, "Initial");
    inputBoard(&Goal, "Goal");
    system("clear");

    // Print Initial and Goal
    printBoard(Initial, "Initial");
    printBoard(Goal, "Goal");

    //Solve
    printHorizontalRule();
    
    solvePuzzle(Initial, Goal);
    return 0;
}

File: Academics/BFS/8PuzzleBFS.js
const readline = require('readline');

const MAXNODES = 1000;
const HorizontalRule = 30;

class GameState {
    constructor(board, blank) {
        this.board = board;
        this.blank = blank;
    }
}

class QueueNode {
    constructor(state, previous, move, depth) {
        this.state = state;
        this.previous = previous;
        this.move = move;
        this.depth = depth;
    }
}

let front = 0, rear = 0, vIndex = 0;
const queue = new Array(MAXNODES);
const visited = new Array(MAXNODES);

function isSameState(state1, state2) {
    for (let i = 0; i < state1.board.length; i++) {
        for (let j = 0; j < state1.board[i].length; j++) {
            if (state1.board[i][j] !== state2.board[i][j]) {
                return false;
            }
        }
    }
    return true;
}

function isVisited(current, visited) {
    for (let i = 0; i < vIndex; i++) {
        if (isSameState(current, visited[i].state)) {
            return true;
        }
    }
    return false;
}

function backtrace(node) {
    if (node === null) return;
    backtrace(node.previous);
    console.log(`Depth ${node.depth} [Move ${node.move}]`);
    printBoard(node.state);
}

function isSafe(x, y, size) {
    return x >= 0 && x < size && y >= 0 && y < size;
}

function swap(board, x1, y1, x2, y2) {
    const temp = board[x1][y1];
    board[x1][y1] = board[x2][y2];
    board[x2][y2] = temp;
}

function printBoard(state) {
    for (let i = 0; i < state.board.length; i++) {
        console.log(state.board[i].join(' '));
    }
    console.log('');
}

function inputBoard(rl, prompt, callback) {
    const size = 3;
    const board = [];
    let blank = { x: 0, y: 0 };
    console.log(`Enter for ${prompt}:\n`);
    function inputElement(i, j) {
        if (i >= size) {
            return callback(new GameState(board, blank));
        }
        if (j >= size) {
            return inputElement(i + 1, 0);
        }

        rl.question(`Enter element for position [${i}][${j}]: `, (input) => {
            const element = parseInt(input);
            if (!board[i]) board[i] = [];
            board[i][j] = element;
            if (element === 0) {
                blank = { x: i, y: j };
            }
            inputElement(i, j + 1);
        });
    }

    inputElement(0, 0);
}

function solvePuzzle(initial, goal) {
    queue[rear] = new QueueNode(initial, null, "Initial State Inclusion", 0);
    rear++;

    while (front < rear && rear < MAXNODES) {
        const current = queue[front].state;
        visited[vIndex] = new QueueNode(current, queue[front].previous, queue[front].move, queue[front].depth);

        const depth = queue[front].depth;
        front++;
        vIndex++;

        if (isSameState(current, goal)) {
            console.log("\nBacktrace Path:\n");
            backtrace(visited[vIndex - 1]);
            console.log(`\nSolution Found at depth ${depth}\n\n`);
            return;
        }

        const { x, y } = current.blank;
        const dy = [-1, 1, 0, 0];
        const dx = [0, 0, -1, 1];
        const moves = ["Left", "Right", "Up", "Down"];

        for (let i = 0; i < 4; i++) {
            const newx = x + dx[i];
            const newy = y + dy[i];

            if (isSafe(newx, newy, current.board.length)) {
                const newState = new GameState(current.board.map(row => row.slice()), { x: newx, y: newy });
                swap(newState.board, x, y, newx, newy);

                if (!isVisited(newState, visited)) {
                    queue[rear] = new QueueNode(newState, visited[vIndex - 1], moves[i], depth + 1);
                    rear++;
                }
            }
        }
    }

    console.log("\nSolution Not Found within Limits.\n\n");
}

function main() {
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
    });

    inputBoard(rl, 'Initial', (initial) => {
        console.clear();
        printBoard(initial);
        inputBoard(rl, 'Goal', (goal) => {
            console.clear();
            printBoard(initial);
            printBoard(goal);
            for (let i = 0; i < HorizontalRule; i++) {
                process.stdout.write("[]");
            }
            solvePuzzle(initial, goal);
            rl.close();
        });
    });
}

// Call the main function to run the program
main();
